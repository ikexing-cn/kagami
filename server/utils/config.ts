import type { Storage } from 'unstorage'
import { createStorage } from 'unstorage'
import fsDriver from 'unstorage/drivers/fs'

let _config: Storage | null = null

const configured = {
  OPENAI_API_KEY: process.env.OPENAI_API_KEY || '',
  OPENAI_API_BASE_URL: process.env.OPENAI_API_BASE_URL || '',
}

export function useConfig() {
  function initialConfig() {
    if (!_config) _config = createStorage({ driver: fsDriver({ base: './config' }) })
  }
  initialConfig()

  const config = _config!

  async function getItem(key: keyof typeof configured) {
    return await config.getItem(key) as string || configured[key]
  }

  function setItem(key: keyof typeof configured, value: string) {
    config.setItem(key, value)
  }

  function getKeys() {
    return Object.keys(configured)
  }

  return {
    getItem,
    setItem,
    getKeys,
  }
}
