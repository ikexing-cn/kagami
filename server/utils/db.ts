import { PrismaClient } from '@prisma/client'

let _prisma: PrismaClient | null = null
export function usePrisma() {
  if (!_prisma)
    _prisma = new PrismaClient()

  return _prisma
}
