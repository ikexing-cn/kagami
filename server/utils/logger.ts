import type { Logger } from 'winston'
import { createLogger, format, transports } from 'winston'
import DailyRotateFile from 'winston-daily-rotate-file'

let _logger: Logger | null = null

export function useLogger(): Logger {
  function cretaeLoggerInner() {
    const _format = format.printf((info) => {
      return `${info.timestamp} [KAgami] ${info.level}: ${info.message}`
    })

    const dailyRotateFileTransport = new DailyRotateFile({
      filename: 'logs/%DATE%.log',
      datePattern: 'YYYY-MM-DD',
      zippedArchive: true,
      maxSize: '10m',
      maxFiles: '14d',
    })

    return createLogger({
      format: format.combine(format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }), _format),
      transports: [
        new transports.Console(),
        dailyRotateFileTransport,
      ],
    })
  }

  if (!_logger)
    _logger = cretaeLoggerInner()

  return _logger
}
