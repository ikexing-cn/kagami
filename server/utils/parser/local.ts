const forwardToken = '[转]'
const onlyNumber = /\[[0-9]+\]/
const fileSize = /\[[0-9.MBmbIi]+\]/
const videoResolution = /\[[0-9]{2,5}(p|P)([A-Za-z0-9]+)?\]/
const fileExtension = ['mp4', 'avi', 'flv', 'mov', 'ts', 'mpg', 'mpeg', 'm2ts', 'rmvb', 'wmv', 'webm', 'rm', 'mkv']
const languageToken = ['jis', 'big5', 'gb', 'cht', 'chs', '简繁', '简日', '简', '繁', '内封', '内嵌', '双语']
const origin = ['web', 'webdl', 'web-dl', 'dl', 'baha', 'bilibili', 'bdrip', 'bd-rip', 'rip']

function parseFileName(name: string, rule?: {}): Record<string, string> {
  name = name.replaceAll('【', '[').replaceAll('】', ']')
  const strs: Array<string> = []
  let pos1 = -1
  let pos2 = -1
  let second = false
  let pair = false

  for (let i = 0; i < name.length; ++i) {
    const c = name[i]
    if (c === '[' || c === ']') {
      if (!second) {
        second = true
        pos1 = i
      } else {
        second = false
        pos2 = i
        pair = true
      }

      if (pair) {
        pair = false
        strs.push(name.substring(pos1, pos2 + 1))
      }
    }
  }

  strs.forEach((v) => {
    name = name.replace(v, '')
  })

  strs.push(name.trim())

  return guessType(strs)
}

function guessType(arr: Array<string>): Record<string, string> {
  const result = {} as Record<string, string>
  const processed = [] as Array<number>
  let pl = 0
  let pr = arr.length - 2
  // skip or read forward token
  if (arr[pl] === forwardToken) {
    result['meta/forward'] = 'true'
    pl++
  }
  // the first bracket is publisher name
  result['meta/publisher-name'] = arr[pl].trim().replace('[', '').replace(']', '')
  // the last bracket usually is file size
  if (fileSize.test(arr[pr])) {
    result['file/size'] = arr[pr]
    processed.push(pr)
    pr--
  }
  const cpr = pr
  for (pr; pr >= pl; --pr) {
    for (const ext of fileExtension) {
      if (arr[pr].toLowerCase() === `[${ext}]`) {
        result['file/type'] = arr[pr].substring(1, arr[pr].length - 1)
        processed.push(pr)
        break
      }
    }
  }
  pr = cpr
  for (pr; pr >= pl; --pr) {
    if (videoResolution.test(arr[pr])) {
      result['meta/video-resolution'] = arr[pr].substring(1, arr[pr].length - 1)
      processed.push(pr)
      break
    }
  }
  pr = cpr
  for (pr; pr >= pl; --pr) {
    // usually is episode
    if (onlyNumber.test(arr[pr])) {
      result['info/episode'] = arr[pr].substring(1, arr[pr].length - 1)
      processed.push(pr)
      break
    }
  }
  pr = cpr
  for (pr; pr >= pl; --pr) {
    for (const ext of languageToken) {
      if (arr[pr].toLowerCase().includes(ext)) {
        result['file/language-propmt'] = arr[pr].substring(1, arr[pr].length - 1)
        processed.push(pr)
        break
      }
    }
  }
  pr = cpr
  for (pr; pr >= pl; --pr) {
    for (const ext of origin) {
      if (arr[pr].toLowerCase().includes(ext)) {
        result['meta/origin'] = arr[pr].substring(1, arr[pr].length - 1)
        processed.push(pr)
        break
      }
    }
  }
  pr = cpr

  return result
}

export { parseFileName }
