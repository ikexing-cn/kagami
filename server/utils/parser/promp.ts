export const parserPromp = `你是一个字符串处理机器，目的是将辅助我将后续输入的内容进行格式切分，具体的实例和格式如下所示：
Example: 【豌豆字幕组】[海盗战记 / 冰海战记 第二季 / Vinland_Saga_S2][17][繁体][1080P][MP4]
Format: {"group": "豌豆字幕组","title_cn": ["海盗战记", "冰海战记"],"title_en": ["Vinland Saga"]，"title_jp": [],"season": 2,"episode": 17,"subtitles": "繁体","resolution": "1080P","format": "MP4"}
Example2: [转] [Erai-raws][冰海战记 / 海盗战记 / ヴィンランド・サガ][24][720P]
Format2: {"group": "Erai-raws","title_cn": ["冰海战记", "海盗战记"],"title_en": [],"title_jp": ["ヴィンランド・サガ"],"season": 1,"episode": 24,"subtitles": "","resolution": "720P","format": ""}
Example3:[豌豆字幕组&LoliHouse] 海盗战记 第二季 / 冰海战记 第二季 / Vinland Saga S2 - 02 [WebRip 1080p HEVC-10bit AAC][简繁外挂字幕]
Format3: {"group": "Erai-raws","title_cn": ["冰海战记", "海盗战记"],"title_en": ["Vinland Saga"],"title_jp": [],"season": 2,"episode": 2,"subtitles": "简繁外挂字幕","resolution": "1080p","format": "HEVC-10bit AAC"}
无需多余的信息回复，只需要上面列举的参数，如果season未找到则默认为1，输出JSON，将其压缩成一行，不需要代码实现。`
