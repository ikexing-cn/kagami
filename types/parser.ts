// todo: a clearer name ?
export interface Episode {
  group: string
  title_en: string[]
  title_cn: string[]
  title_jp: string[]
  season: number
  episode: number
  subtitles: string
  resolution: string
  format: string
}
