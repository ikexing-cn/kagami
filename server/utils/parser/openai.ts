import { ChatGPTAPI } from 'chatgpt'
import { useConfig } from '../config'
import { parserPromp } from './promp'

let _chatGPT: InstanceType<typeof ChatGPTAPI> | null = null

export async function useChatGPT() {
  if (!_chatGPT) {
    const { getItem } = useConfig()

    _chatGPT = new ChatGPTAPI({
      apiKey: await getItem('OPENAI_API_KEY'),
      apiBaseUrl: await getItem('OPENAI_API_BASE_URL'),
      systemMessage: parserPromp,
    })
  }

  return _chatGPT
}

export async function useChatGPTParse(msg: string) {
  const chatgpt = await useChatGPT()
  return JSON.parse((await chatgpt.sendMessage(msg)).text)
}
