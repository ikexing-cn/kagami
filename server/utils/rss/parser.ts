import { parse as anitomyParser } from 'anitomy'
import { XMLParser } from 'fast-xml-parser'
import { objectPick } from '@vueuse/core'
import { useChatGPTParse } from '../parser/openai'
import type { Episode } from '~/types/parser'

interface Item {
  guid: string
  link: string
  title: string
  description: string
  torrent: Torrent
  enclosure: string

  episode: number
}

interface Torrent {
  link: string
  contentLength: number
  pubDate: Date
}

const simolifedChinese = ['GB', 'CHS', 'SC', '简体']
const traditionalChinese = ['BIG5', 'CHT', 'TC', '繁体', '简繁']

function getEpisodeByAnitomy(title: string) {
  // todo: handle maybe undefined at the parser
  return anitomyParser(title)?.episode?.number ?? 0
}

function calculatePriority(title: string, episode: number, _options?: {
  useSimolifedChinese?: boolean
  useEmbeddedSubtitles?: boolean
}) {
  function reverse(str: string) {
    return str.split('').reverse().join('')
  }

  const titleReversal = reverse(title)
  const options = {
    useSimolifedChinese: true,
    useEmbeddedSubtitles: false,
    ..._options,
  }

  let isSimolifed = false
  for (const item of simolifedChinese) {
    if (titleReversal.includes(reverse(item))) {
      isSimolifed = true
      break
    }
  }

  let isTraditional = false
  if (!isSimolifed) {
    for (const item of traditionalChinese) {
      if (titleReversal.includes(reverse(item))) {
        isTraditional = true
        break
      }
    }
  }

  let result = episode
  if (options.useSimolifedChinese) {
    if (isSimolifed)
      result += 200
  } else if (isTraditional) {
    result += 200
  }

  if (options.useEmbeddedSubtitles) {
    if (title.includes('内嵌'))
      result += 100
  } else if (title.includes('内封')) {
    result += 100
  }

  return result
}

export async function createRssJson(url: string) {
  const xmlResult = await (await fetch(url)).text()
  // nerver trust other fields in the `rss.channel`
  const jsonResultItems = new XMLParser().parse(xmlResult).rss.channel.item as Item[]

  const sortedItems = jsonResultItems.map((item) => {
    item.episode = getEpisodeByAnitomy(item.title)
    return item
  }).sort((a, b) => a.episode - b.episode)
    .map(item => objectPick(item, ['title', 'link', 'episode', 'torrent']))
    // todo: priority options
    .sort((a, b) => calculatePriority(b.title, b.episode) - calculatePriority(a.title, a.episode))

  const parserResult = await useChatGPTParse(sortedItems[0].title) as Episode

  return {
    formatExample: parserResult,
    items: sortedItems.splice(0, sortedItems[0].episode),
  }
}
