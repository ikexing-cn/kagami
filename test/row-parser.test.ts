import * as dotenv from 'dotenv'
import { describe, expect, it } from 'vitest'
import { useChatGPTParse } from '../server/utils/parser/openai'
import { useLogger } from '../server/utils/logger'
import { parseFileName } from '../server/utils/parser/local'
import type { Episode } from '~/types/parser'

dotenv.config()
const logger = useLogger()

describe.skip('chatgpt-row-parser', async () => {
  it('gpt parser', async () => {
    const { text } = await useChatGPTParse('[Lilith-Raws] 和山田谈场 Lv999 的恋爱 / Yamada-kun to Lv999 no Koi wo Suru - 03 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]')
    const res = JSON.parse(text) as Episode
    expect(res).toMatchObject({
      group: 'Lilith-Raws',
      title_cn: ['和山田谈场 Lv999 的恋爱'],
      title_en: ['Yamada-kun to Lv999 no Koi wo Suru'],
      title_jp: [],
      season: 1,
      episode: 3,
      subtitles: 'CHT',
      resolution: '1080p',
      // skip format because gpt may will output `AVC AAC MP4` or `AVC AAC`
    })
  })
})

describe('local parser', async () => {
  it('local parser', () => {
    const text = '[爱恋字幕社][熊熊勇闯异世界 Punch!][Kuma Kuma Kuma Bear Punch!][05][1080P][MP4][GB][简中] [292.23 MB]'
    const value = parseFileName(text)
    for (const v in value)
      logger.info(`${v}: ${value[v]}`)
  })
})
