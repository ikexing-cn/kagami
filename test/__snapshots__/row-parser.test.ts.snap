// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`local parser > local parser 1`] = `
{
  "file/type": "[MP4]",
  "meta/publisher-name": "[Lilith-Raws]",
}
`;
